protocol/directory.go

```diff
@@ -54,56 +54,86 @@ func (d *ConiksDirectory) LatestSTR() *merkletree.SignedTreeRoot {
 	return d.pad.LatestSTR()
 }
 
-func (d *ConiksDirectory) Register(uname string, key []byte) (
-	*merkletree.AuthenticationPath, *merkletree.TemporaryBinding, ErrorCode) {
+func (d *ConiksDirectory) Register(req *RegistrationRequest) (
+	*RegistrationResponse, ErrorCode) {
 	// check whether the name already exists
 	// in the directory before we register
-	ap, err := d.pad.Lookup(uname)
+	ap, err := d.pad.Lookup(req.Username)
 	if err != nil {
-		return nil, nil, ErrorInternalServer
+		return nil, ErrorDirectory
 	}
 	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index()) {
-		return ap, nil, ErrorNameExisted
+		return &RegistrationResponse{
+			Type: RegistrationType,
+			AP:   ap,
+			STR:  d.LatestSTR(),
+			TB:   nil,
+		}, ErrorNameExisted
 	}
 	if d.useTBs {
 		// also check the temporary bindings array
 		// currently the server allows only one registration/key change per epoch
-		if tb := d.tbs[uname]; tb != nil {
-			return ap, tb, ErrorNameExisted
+		if tb := d.tbs[req.Username]; tb != nil {
+			return &RegistrationResponse{
+				Type: RegistrationType,
+				AP:   ap,
+				STR:  d.LatestSTR(),
+				TB:   tb,
+			}, ErrorNameExisted
 		}
 
 		// insert new data to the directory on-the-fly
-		tb, err := d.pad.TB(uname, key)
+		tb, err := d.pad.TB(req.Username, req.Key)
 		if err != nil {
-			return nil, nil, ErrorInternalServer
+			return nil, ErrorDirectory
 		}
-		d.tbs[uname] = tb
-		return ap, tb, Success
-	}
-
-	if err = d.pad.Set(uname, key); err != nil {
-		return nil, nil, ErrorInternalServer
+		d.tbs[req.Username] = tb
+		return &RegistrationResponse{
+			Type: RegistrationType,
+			AP:   ap,
+			STR:  d.LatestSTR(),
+			TB:   tb,
+		}, Success
+	} else {
+		if err = d.pad.Set(req.Username, req.Key); err != nil {
+			return nil, ErrorDirectory
+		}
+		return &RegistrationResponse{
+			Type: RegistrationType,
+			AP:   ap,
+			STR:  d.LatestSTR(),
+			TB:   nil,
+		}, Success
 	}
-
-	return ap, nil, Success
 }
 
-func (d *ConiksDirectory) KeyLookup(uname string) (
-	*merkletree.AuthenticationPath, *merkletree.TemporaryBinding, ErrorCode) {
-	ap, err := d.pad.Lookup(uname)
+func (d *ConiksDirectory) KeyLookup(req *KeyLookupRequest) (
+	*KeyLookupResponse, ErrorCode) {
+	ap, err := d.pad.Lookup(req.Username)
 	if err != nil {
-		return nil, nil, ErrorInternalServer
+		return nil, ErrorDirectory
 	}
 
 	if !bytes.Equal(ap.LookupIndex, ap.Leaf.Index()) {
 		// if not found in the tree, do lookup in tb array
 		if d.useTBs {
-			if tb := d.tbs[uname]; tb != nil {
-				return ap, tb, Success
+			if tb := d.tbs[req.Username]; tb != nil {
+				return &KeyLookupResponse{
```

Comment:
Similarly here, I think this should be replaced with a setter to make the code a bit more concise.